Attempt 2 - Adds player inventory interaction / stripping

chairbender opened the pull request:
### Purpose
Removed the broken attempted tab update fixes.

Fixes #2144 

Also the "Networking > Respawn Player" command should work now regardless of game mode.

Can click drag a non-conscious player to yourself to view their inventory. A button allows dropping all of their items on the ground, or you can interact with their inventory slots just like it was a storage you are looking into.

It's a bit "quick and dirty" as far as the UI goes but at least now you can do it and items aren't lost forever in their inventory.

### Open Questions and Pre-Merge TODOs

- [X]  This fix is tested on the same branch it is PR'ed to.
- [X]  I correctly commented my code
- [X]  My code is indented with tabs and not spaces
- [X]  This PR does not include any unnecessary .meta, .prefab or .unity (scene) changes
- [X]  This PR does not bring up any new compile errors
- [X]  This PR has been tested in editor
- [X]  Any new / changed components have been [tested using Respawn](https://github.com/unitystation/unitystation/wiki/Proper-Spawning-and-Despawning-%28Object-Lifecycle%29)
- [ ]  This PR has been tested with round restarts.
- [X]  This PR has been tested on moving / rotating / rotated-before-joining matrices (if applicable)
- [X]  This PR has been tested in multiplayer (with 2 clients and late joiners, if applicable)

chairbender committed:
fix #2251 and make it so that large tab update / open
messages are broken up

chairbender committed:
#2144 WIP adding ability to interact with other players' inventory

chairbender committed:
#2144 get other player inventory interaction working server side

chairbender committed:
Revert "fix #2251 and make it so that large tab update / open messages are broken up"

This reverts commit 1ae39e24

chairbender committed:
#2144 more robust player inventory interaction, and remove
accidentally added Pickupable to player. Adds event hooks so
storage will be closed when player becomes no longer observable

chairbender committed:
#2144 even more robust player inventory interactions.
Add hand slots to other player inventory display.
Drop items when cuffed.
Don't allow interactions when cuffed.
Fix unrelated NRE that happened on server spawning while a client
was joining.

fomalsd commented on UnityProject/Assets/Scripts/Input System/InteractionV2/Validations.cs:
```diff
@@ -350,4 +332,26 @@ public static bool CanFit(ItemSlot itemSlot, Pickupable toCheck, NetworkSide sid
+		else
+		{
+			//find their exact conscious state, slipping state, cuffed state
+			var playerHealth = player.GetComponent<PlayerHealth>();
```
Could reduce `GetComponent` calls here by getting just the `PlayerScript`, it has references to all the scripts you're trying to get here

fomalsd commented on UnityProject/Assets/Scripts/Player/PlayerNetworkActions.cs:
```diff
@@ -81,6 +81,28 @@ public void CmdDropItem(NamedSlot equipSlot)
+		{
+			//skip slots which have special uses
+			if (itemSlot.NamedSlot == NamedSlot.handcuffs) continue;
+			Inventory.ServerDrop(itemSlot);
```
would be nice to convert this into a progress action with a coroutine down the road, taking like .2s for each emptied slot 
as this "insta naked" button seems a bit overpowered at the moment

chairbender replied:
I'll do this. It should be easy with the new Progress action system.

fomalsd commented on UnityProject/Assets/Scripts/Tilemaps/Behaviours/Objects/SpatialRelationship/ObserveStorageRelationship.cs:
```diff
@@ -0,0 +1,92 @@
+		//stop the relationship if observed player becomes conscious
+		if (newState == ConsciousState.CONSCIOUS)
+		{
+			SpatialRelationship.ServerEnd(this);
```
shouldn't this check other factors sustaining relationship before breaking it? like, if player is cuffed all the time it should probably remain unbroken regardless of conscious/slipped state change
but if you intentionally skipped on this to avoid code getting overly complex then it's understandable

chairbender replied:
yeah I thought it was just easier for now to just break it once any factor changes, but actually the logic you're talking about should be quite simple so I'll add that

chairbender committed:
#2144 add interaction validations to all appropriate
 PNA / WNA Cmds so they can't be done while cuffed. Re-enable
 dummy spawning. Modify logic for ObserveStorageRelationship so you can
 keep looking as long as at least one of the criteria are met. Refactor
 SpatialRelationship slightly to improve naming / support these use
 cases.

chairbender committed:
Remove canNotInteract and change ALL validations that were relying on it
to use Validations.CanInteract or Validations.CanApply as the
new standards, to centralize interaction validation logic. Fix various
bugs with cuffing / stripping, add progress action for strip all
command.

fomalsd reviewed (approved):
awesome
